Modules/blake2module.c

```diff
@@ -0,0 +1,924 @@
+    int eax7 = 0, ebx7 = 0, ecx7 = 0, edx7 = 0;
+#if defined(__x86_64__) && defined(__GNUC__)
+    __cpuid_count(1, 0, eax1, ebx1, ecx1, edx1);
+    __cpuid_count(7, 0, eax7, ebx7, ecx7, edx7);
+#elif defined(_M_X64)
+    int info1[4] = { 0 };
+    int info7[4] = { 0 };
+    __cpuidex(info1, 1, 0);
+    __cpuidex(info7, 7, 0);
+    eax1 = info1[0];
+    ebx1 = info1[1];
+    ecx1 = info1[2];
+    edx1 = info1[3];
+    eax7 = info7[0];
+    ebx7 = info7[1];
+    ecx7 = info7[2];
+    edx7 = info7[3];
+#else
+    (void) eax1; (void) ebx1; (void) ecx1; (void) edx1;
+    (void) eax7; (void) ebx7; (void) ecx7; (void) edx7;
+#endif
+
+    avx = (ecx1 & ECX_AVX) != 0;
+
+    avx2 = (ebx7 & EBX_AVX2) != 0;
+
+    sse = (edx1 & EDX_SSE) != 0;
+    sse2 = (edx1 & EDX_SSE2) != 0;
+    cmov = (edx1 & EDX_CMOV) != 0;
+
+    sse3 = (ecx1 & ECX_SSE3) != 0;
+    /* ssse3 = (ecx1 & ECX_SSSE3) != 0; */
+    sse41 = (ecx1 & ECX_SSE4_1) != 0;
+    sse42 = (ecx1 & ECX_SSE4_2) != 0;
+
+    done = true;
+  }
+}
+
+static inline bool has_simd128(void) {
+  // For now this is Intel-only, could conceivably be #ifdef'd to something
+  // else.
+  return sse && sse2 && sse3 && sse41 && sse42 && cmov;
+}
+
+static inline bool has_simd256(void) {
+  return avx && avx2;
+}
+
+// Small mismatch between the variable names Python defines as part of configure
+// at the ones HACL* expects to be set in order to enable those headers.
+#define HACL_CAN_COMPILE_VEC128 HACL_CAN_COMPILE_SIMD128
+#define HACL_CAN_COMPILE_VEC256 HACL_CAN_COMPILE_SIMD256
+
+#include "_hacl/Hacl_Hash_Blake2b.h"
+#include "_hacl/Hacl_Hash_Blake2s.h"
+#if HACL_CAN_COMPILE_SIMD256
+#include "_hacl/Hacl_Hash_Blake2b_Simd256.h"
+#endif
+#if HACL_CAN_COMPILE_SIMD128
+#include "_hacl/Hacl_Hash_Blake2s_Simd128.h"
+#endif
+
+// MODULE TYPE SLOTS
+
+static PyType_Spec blake2b_type_spec;
+static PyType_Spec blake2s_type_spec;
+
+PyDoc_STRVAR(blake2mod__doc__,
+"_blake2b provides BLAKE2b for hashlib\n"
+);
+
+typedef struct {
+    PyTypeObject* blake2b_type;
+    PyTypeObject* blake2s_type;
+} Blake2State;
+
+static inline Blake2State*
+blake2_get_state(PyObject *module)
+{
+    void *state = PyModule_GetState(module);
+    assert(state != NULL);
+    return (Blake2State *)state;
+}
+
+static struct PyMethodDef blake2mod_functions[] = {
+    {NULL, NULL}
+};
+
+static int
+_blake2_traverse(PyObject *module, visitproc visit, void *arg)
+{
+    Blake2State *state = blake2_get_state(module);
+    Py_VISIT(state->blake2b_type);
+    Py_VISIT(state->blake2s_type);
+    return 0;
+}
+
+static int
+_blake2_clear(PyObject *module)
+{
+    Blake2State *state = blake2_get_state(module);
+    Py_CLEAR(state->blake2b_type);
+    Py_CLEAR(state->blake2s_type);
+    return 0;
+}
+
+static void
+_blake2_free(void *module)
+{
+    _blake2_clear((PyObject *)module);
+}
+
+#define ADD_INT(d, name, value) do { \
+    PyObject *x = PyLong_FromLong(value); \
+    if (!x) \
+        return -1; \
```

Comment:
```suggestion
    if (x == NULL) { \
        return -1; \
    } \
```

(I'll let you align the backslashes)